packages/wrangler/src/experimental-assets.ts

```diff
@@ -1,8 +1,299 @@
+					await new Promise((resolvePromise) =>
+						setTimeout(resolvePromise, Math.pow(2, attempts) * 1000)
+					);
+					// TODO: handle other errors e.g. jwt expired here
+					if (e instanceof APIError && e.isGatewayError()) {
+						// Gateway problem, wait for some additional time and set concurrency to 1
+						queue.concurrency = 1;
+						await new Promise((resolvePromise) =>
+							setTimeout(resolvePromise, Math.pow(2, gatewayErrors) * 5000)
+						);
+
+						gatewayErrors++;
+
+						// only count as a failed attempt after a few initial gateway errors
+						if (gatewayErrors >= MAX_UPLOAD_GATEWAY_ERRORS) {
+							attempts++;
+						}
+					} else {
+						attempts++;
+					}
+					return doUpload();
+				} else {
+					logger.debug("failed:", e);
+					throw e;
+				}
+			}
+		};
+		// add to queue and run it if we haven't reached concurrency limit
+		void queue.add(() =>
+			doUpload().then(
+				(res) => {
+					bucketUploadCount++;
+					completionJwt = res.jwt || completionJwt;
+				},
+				(error) => {
+					return Promise.reject(
+						new FatalError(
+							`File upload failed. Please try again. Error: ${JSON.stringify(
+								error
+							)})`,
+							1
+						)
+					);
+				}
+			)
+		);
+	}
+	await queue.onIdle();
+	if (!completionJwt) {
+		throw new FatalError("Failed to upload all files. Please try again.", 1);
+	} else if (bucketUploadCount !== initializeAssetsResponse.buckets.length) {
+		throw new FatalError(
+			"Completion signal received unexpectedly early - we cannot confirm if all files were successfully uploaded. Please try again.",
+			1
+		);
+	}
+
+	const uploadMs = Date.now() - start;
+	const skipped = Object.keys(manifest).length - includedHashes.length;
+	const skippedMessage = skipped > 0 ? `(${skipped} already uploaded) ` : "";
+
+	logger.log(
+		`✨ Success! Uploaded ${
+			filteredFiles.length
+		} files ${skippedMessage}${formatTime(uploadMs)}\n`
+	);
+
+	return { manifest, jwt: completionJwt };
+};
+
+// modified from /pages/validate.tsx
+const walk = async (
+	dir: string,
+	manifest: AssetManifest,
+	startingDir: string = dir
+) => {
+	const files = await readdir(dir);
+
+	let counter = 0;
+	await Promise.all(
+		files.map(async (file) => {
+			const filepath = path.join(dir, file);
+			const relativeFilepath = path.relative(startingDir, filepath);
+			const filestat = await stat(filepath);
+
+			if (filestat.isSymbolicLink()) {
+				return;
+			}
+
+			if (filestat.isDirectory()) {
+				manifest = await walk(filepath, manifest, startingDir);
+			} else {
+				if (counter >= MAX_ASSET_COUNT) {
+					throw new UserError(
+						`You cannot have more than ${MAX_ASSET_COUNT.toLocaleString()} files in a deployment. Ensure you have specified your build output directory correctly.`
+					);
+				}
+
+				const name = urlSafe(relativeFilepath);
+				if (filestat.size > MAX_ASSET_SIZE) {
+					throw new FatalError(
+						`Max file size is ${prettyBytes(MAX_ASSET_SIZE, {
```

Comment:
similar to https://github.com/cloudflare/workers-sdk/pull/6374/files#r1701820626